x-pack/plugins/security_solution/server/lib/detection_engine/rule_types/factories/utils/strip_non_ecs_fields.ts

```diff
@@ -0,0 +1,103 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License
+ * 2.0; you may not use this file except in compliance with the Elastic License
+ * 2.0.
+ */
+
+import { ecsFieldMap } from '@kbn/rule-registry-plugin/common/assets/field_maps/ecs_field_map';
+
+import { isPlainObject, cloneDeep, unset } from 'lodash';
+
+import type { SearchTypes } from '../../../../../../common/detection_engine/types';
+
+// const getECSObjectFields =
+
+/**
+ * retrieve all nested object fields from ecsFieldMap
+ * field `agent.build.original` will be converted into
+ * { agent: true, agent.build: true }
+ */
+const getEcsObjectFields = () => {
+  const result: Record<string, boolean> = {};
+
+  Object.entries(ecsFieldMap).forEach(([key, value]) => {
+    const objects = key.split('.');
+    // last item can be any of type, as precedent are objects
+    objects.pop();
+
+    objects.reduce((parentPath, itemKey) => {
+      const fullPath = parentPath ? `${parentPath}.${itemKey}` : itemKey;
+
+      if (!result[fullPath]) {
+        result[fullPath] = true;
+      }
+
+      return fullPath;
+    }, '');
+  });
+
+  return result;
+};
+
+const ecsObjectFields = getEcsObjectFields();
+
+/**
+ * check whether field is value is ECS compliant
+ */
+const computeIsEcsCompliant = (value: Record<string, SearchTypes> | SearchTypes, path: string) => {
+  const ecsField = ecsFieldMap[path as keyof typeof ecsFieldMap];
+
+  const isEcsFieldObject =
+    ['object', 'flattened'].includes(ecsField?.type) || ecsObjectFields[path];
+
+  // if checked value is object but ECS field not, it's not compliant
+  if (isPlainObject(value) && !isEcsFieldObject) {
+    return false;
+  }
+
+  // if checked value is not object but ECS field is, it's not compliant
+  if (!isPlainObject(value) && isEcsFieldObject) {
+    return false;
+  }
+
+  return true;
```

Comment:
Another edge case involves dot-notation fields in the source documents: it's possible to define a field like `'agent.name.conflict': 'some-value'` in a source document with the whole field path a single string, which would cause us to skip past the check where we would identify `agent.name` as a conflict in this logic.

```
it('should strip conflicting fields that use dot notation', () => {
    expect(
      stripNonEcsFields({
        'agent.name.conflict': 'some-value',
        message: 'test message',
      })
    ).toEqual({
      result: {
        message: 'test message',
      },
      removed: [
        {
          key: 'agent.name.conflict',
          value: 'some-value',
        },
      ],
    });
  });
```